src/tuple_impl.rs

```diff
@@ -236,8 +238,29 @@ where
     type Item = T;
 
     fn next(&mut self) -> Option<Self::Item> {
+        self.len = self.len.saturating_sub(1);
         self.iter.next()
     }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        (self.len, Some(self.len))
```

Comment:
Rather than add an additional field, why not just forward the underlying implementation of `size_hint`. 
```suggestion
        self.iter.size_hint()
```

Reply:
The problem is the type of the `iter` field: `Take<TupleWindows<Cycle<I>, T>>`.
Even if `Take::size_hint` does a good job, `TupleWindows` does not even have access to the original iterator as it's wrapped by `Cycle` and `TupleWindows` won't store the length at definition as it does not require `I` to be `ExactSizeIterator`.
I guess we could rewrite `CircularTupleWindows` entirely to not rely on such wrapped type, but I'm not very keen on that to save a field.

Reply:
Ahh.. maybe we can ensure that `TupleWindows<Cycle<I>, T>` has an overflowing size hint then `Take` would successfully do the job.

Reply:
As I feared, it's not possible in some cases. `TupleWindows` would have fewer items.

    fn size_hint(&self) -> (usize, Option<usize>) { // TupleWindows
        size_hint::sub_scalar(self.iter.size_hint(), T::num_items() - 1)
    }
    // Might still be valuable to commit thought.

The test below would fail

    [(); usize::MAX].iter().circular_tuple_windows::<(_, _, _, _)>().len() == usize::MAX

However it's not a problem, I'm gonna unwrap `Take` type as it's just `struct Take<I> { iter: I, n: usize }`.